File: 07550.py
import sys
import string


sys.setrecursionlimit(10 ** 6)


def parse(line):
    words = []
    word = ""
    for c in line:
        if c in string.ascii_letters:
            word += c
        else:
            if word:
                words.append(word)
            word = ""
    if word:
        words.append(word)
    return words


def final_word(dest, words, threshold, idx):
    if threshold <= idx < len(words):
        return words[idx]
    elif idx >= len(words):
        return "-outside-"
    elif idx in dest:
        return dest[idx]
    dest[idx] = final_word(dest, words, threshold, idx + len(words[idx]))
    return dest[idx]


def magic_trick(words1, words2, words3):
    ans = set()
    dest = {}
    words = words1 + words2 + words3
    threshold = len(words1) + len(words2)
    for idx, _ in enumerate(words1):
        ans.add(final_word(dest, words, threshold, idx).lower())
        if len(ans) > 3:
            return ["-too many-"]
    return sorted(ans)


def main():
    t = int(sys.stdin.readline())
    for a0 in range(1, t + 1):
        words1 = parse(sys.stdin.readline().rstrip())
        words2 = parse(sys.stdin.readline().rstrip())
        words3 = parse(sys.stdin.readline().rstrip())
        print(f"Scenario #{a0}:")
        print("\n".join(magic_trick(words1, words2, words3)) + "\n")


main()

File: 07550.cpp
#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <algorithm>
#define MAX 100000

using namespace std;

void parse(string &line, vector<string> &words){
	string word;
	for(char c : line){
		if(c >= 'a' && c <= 'z'){
			word += c;
		}else if(c >= 'A' && c <= 'Z'){
			word += (char)(c - 'A' + 'a');
		}else{
			if(!word.empty()){
				words.push_back(word);
			}
			word = "";
		}
	}
	if(!word.empty()){
		words.push_back(word);
	}
}

unsigned final_word(unsigned dest[MAX], vector<string> &words, unsigned thres, unsigned idx){
	if(idx >= thres && idx < words.size()){
		return idx;
	}else if(idx >= words.size()){
		return 0;
	}else if(dest[idx]){
		return dest[idx];
	}
	dest[idx] = final_word(dest, words, thres, idx + words[idx].length());
	return dest[idx];
}

vector<string> magic_trick(vector<string> &words1, vector<string> &words2, vector<string> &words3){
	set<unsigned> ans;
	unsigned dest[MAX] = {};
	vector<string> words;
	for(string &word : words1){
		words.push_back(word);
	}
	for(string &word : words2){
		words.push_back(word);
	}
	for(string &word : words3){
		words.push_back(word);
	}
	unsigned thres = words1.size() + words2.size();
	vector<string> ret;
	for(unsigned idx = 0; idx < words1.size(); ++idx){
		ans.insert(final_word(dest, words, thres, idx));
		if(ans.size() > 3){
			ret.push_back("-too many-");
			return ret;
		}
	}
	for(unsigned a : ans){
		if(a == 0){
			ret.push_back("-outside-");
		}else{
			ret.push_back(words[a]);
		}
	}
	sort(ret.begin(), ret.end());
	return ret;
}

int main(){
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	int t;
	cin >> t;
	string line;
	getline(cin, line);
	for(int a0 = 1; a0 <= t; ++a0){
		vector<string> words1, words2, words3;
		getline(cin, line);
		parse(line, words1);
		getline(cin, line);
		parse(line, words2);
		getline(cin, line);
		parse(line, words3);
		cout << "Scenario #" << a0 << ":\n";
		vector<string> ans = magic_trick(words1, words2, words3);
		for(string s : ans){
			cout << s << "\n";
		}
		cout << "\n";
	}
	return 0;
}
